### ISSUE TITLE
"unknown error" instead of "element not interactable" on an attempt to sendKeys to an invisible field

### ISSUE THREAD

**Issue by barancev** [mozilla/geckodriver#579]
## Firefox Version
55.0a1 (2017-03-27) (64-bit) on Windows 7
geckodriver 0.15

## Steps to reproduce -
Try to sendKeys to an invisible field.
geckodriver::marionette throws a legit "element not interactable" error
but webdriver::server rethrows it as "unknown error"

```
1490717721324	webdriver::server	DEBUG	Got request POST AbsolutePath("/session/90c29c02-42d9-44ce-8013-cf08ebec1e5f/element/156a6a58-685b-4196-928e-2908e552be9d/value")
1490717721325	webdriver::command	DEBUG	Got request body {"id":"156a6a58-685b-4196-928e-2908e552be9d","text":["Y","o","u"," ","d","o","n","'","t"," ","s","e","e"," ","m","e"],"value":["Y","o","u"," ","d","o","n","'","t"," ","s","e","e"," ","m","e"]}
1490717721325	geckodriver::marionette	DEBUG	→ 145:[0,4,"sendKeysToElement",{"id":"156a6a58-685b-4196-928e-2908e552be9d","value":["Y","o","u"," ","d","o","n","'","t"," ","s","e","e"," ","m","e"]}]
1490717721357	Marionette	TRACE	conn0 <- [1,4,{"error":"element not interactable","message":"Element is not visible","stacktrace":"WebDriverError@chrome://marionette/content/error.js:200:5\nElementNotInteractableError@chrome://marionette/content/error.js:276:5\nevent.sendKeysToElement@chrome://marionette/content/event.js:1310:11\ninteraction.sendKeysToElement/<@chrome://marionette/content/interaction.js:361:5\npromise callback*interaction.sendKeysToElement@chrome://marionette/content/interaction.js:359:10\nsendKeysToElement@chrome://marionette/content/listener.js:1468:11\nTaskImpl_run@resource://gre/modules/Task.jsm:319:42\nTaskImpl@resource://gre/modules/Task.jsm:277:3\nasyncFunction@resource://gre/modules/Task.jsm:252:14\nTask_spawn@resource://gre/modules/Task.jsm:166:12\nTaskImpl_handleResultValue@resource://gre/modules/Task.jsm:389:16\nTaskImpl_run@resource://gre/modules/Task.jsm:327:15\nTaskImpl@resource://gre/modules/Task.jsm:277:3\nasyncFunction@resource://gre/modules/Task.jsm:252:14\nTask_spawn@resource://gre/modules/Task.jsm:166:12\ndispatch/<@chrome://marionette/content/listener.js:186:15\n"},null]
1490717721357	geckodriver::marionette	DEBUG	← [1,4,{"error":"element not interactable","message":"Element is not visible","stacktrace":"WebDriverError@chrome://marionette/content/error.js:200:5\nElementNotInteractableError@chrome://marionette/content/error.js:276:5\nevent.sendKeysToElement@chrome://marionette/content/event.js:1310:11\ninteraction.sendKeysToElement/<@chrome://marionette/content/interaction.js:361:5\npromise callback*interaction.sendKeysToElement@chrome://marionette/content/interaction.js:359:10\nsendKeysToElement@chrome://marionette/content/listener.js:1468:11\nTaskImpl_run@resource://gre/modules/Task.jsm:319:42\nTaskImpl@resource://gre/modules/Task.jsm:277:3\nasyncFunction@resource://gre/modules/Task.jsm:252:14\nTask_spawn@resource://gre/modules/Task.jsm:166:12\nTaskImpl_handleResultValue@resource://gre/modules/Task.jsm:389:16\nTaskImpl_run@resource://gre/modules/Task.jsm:327:15\nTaskImpl@resource://gre/modules/Task.jsm:277:3\nasyncFunction@resource://gre/modules/Task.jsm:252:14\nTask_spawn@resource://gre/modules/Task.jsm:166:12\ndispatch/<@chrome://marionette/content/listener.js:186:15\n"},null]
1490717721358	webdriver::server	DEBUG	Returning status InternalServerError
1490717721358	webdriver::server	DEBUG	Returning body {"value":{"error":"unknown error","message":"Element is not visible","stacktrace":"stack backtrace:\n   0:           0x61f07f - <no info>\n   1:           0x61fe59 - <no info>\n   2:           0x43b0b5 - <no info>\n   3:           0x420313 - <no info>\n   4:           0x408307 - <no info>\n   5:           0x6750f9 - <no info>\n   6:           0x416877 - <no info>\n   7:           0x66ed53 - <no info>\n   8:         0x76b959cd - BaseThreadInitThunk"}}
```

**Comment by barancev** [comment]
Similar issue for a click to an invisible element:

```
1490718526443	webdriver::server	DEBUG	Got request POST AbsolutePath("/session/65fe094b-b150-4879-aeeb-fd279970d7d0/element/f03c9743-46c1-4c4b-8334-f3e622be5a03/click")
1490718526444	webdriver::command	DEBUG	Got request body {"id":"f03c9743-46c1-4c4b-8334-f3e622be5a03"}
1490718526444	geckodriver::marionette	DEBUG	→ 66:[0,4,"clickElement",{"id":"f03c9743-46c1-4c4b-8334-f3e622be5a03"}]
1490718526445	Marionette	TRACE	conn0 -> [0,4,"clickElement",{"id":"f03c9743-46c1-4c4b-8334-f3e622be5a03"}]
1490718526478	Marionette	TRACE	conn0 <- [1,4,{"error":"element not interactable","message":"","stacktrace":"WebDriverError@chrome://marionette/content/error.js:200:5\nElementNotInteractableError@chrome://marionette/content/error.js:276:5\nseleniumClickElement@chrome://marionette/content/interaction.js:204:11\nTaskImpl_run@resource://gre/modules/Task.jsm:319:42\nTaskImpl@resource://gre/modules/Task.jsm:277:3\nasyncFunction@resource://gre/modules/Task.jsm:252:14\nTask_spawn@resource://gre/modules/Task.jsm:166:12\nTaskImpl_handleResultValue@resource://gre/modules/Task.jsm:389:16\nTaskImpl_run@resource://gre/modules/Task.jsm:327:15\nTaskImpl@resource://gre/modules/Task.jsm:277:3\nasyncFunction@resource://gre/modules/Task.jsm:252:14\nTask_spawn@resource://gre/modules/Task.jsm:166:12\nTaskImpl_handleResultValue@resource://gre/modules/Task.jsm:389:16\nTaskImpl_run@resource://gre/modules/Task.jsm:327:15\nTaskImpl@resource://gre/modules/Task.jsm:277:3\nasyncFunction@resource://gre/modules/Task.jsm:252:14\nTask_spawn@resource://gre/modules/Task.jsm:166:12\ndispatch/<@chrome://marionette/content/listener.js:186:15\n"},null]
1490718526478	geckodriver::marionette	DEBUG	← [1,4,{"error":"element not interactable","message":"","stacktrace":"WebDriverError@chrome://marionette/content/error.js:200:5\nElementNotInteractableError@chrome://marionette/content/error.js:276:5\nseleniumClickElement@chrome://marionette/content/interaction.js:204:11\nTaskImpl_run@resource://gre/modules/Task.jsm:319:42\nTaskImpl@resource://gre/modules/Task.jsm:277:3\nasyncFunction@resource://gre/modules/Task.jsm:252:14\nTask_spawn@resource://gre/modules/Task.jsm:166:12\nTaskImpl_handleResultValue@resource://gre/modules/Task.jsm:389:16\nTaskImpl_run@resource://gre/modules/Task.jsm:327:15\nTaskImpl@resource://gre/modules/Task.jsm:277:3\nasyncFunction@resource://gre/modules/Task.jsm:252:14\nTask_spawn@resource://gre/modules/Task.jsm:166:12\nTaskImpl_handleResultValue@resource://gre/modules/Task.jsm:389:16\nTaskImpl_run@resource://gre/modules/Task.jsm:327:15\nTaskImpl@resource://gre/modules/Task.jsm:277:3\nasyncFunction@resource://gre/modules/Task.jsm:252:14\nTask_spawn@resource://gre/modules/Task.jsm:166:12\ndispatch/<@chrome://marionette/content/listener.js:186:15\n"},null]
1490718526479	webdriver::server	DEBUG	Returning status InternalServerError
1490718526479	webdriver::server	DEBUG	Returning body {"value":{"error":"unknown error","message":"","stacktrace":"stack backtrace:\n   0:           0x61f07f - <no info>\n   1:           0x61fe59 - <no info>\n   2:           0x43b0b5 - <no info>\n   3:           0x420313 - <no info>\n   4:           0x408307 - <no info>\n   5:           0x6750f9 - <no info>\n   6:           0x416877 - <no info>\n   7:           0x66ed53 - <no info>\n   8:         0x76b959cd - BaseThreadInitThunk"}}
```

**Comment by andreastt** [comment]
Thanks for reporting!

It looks like the webdriver crate is still using the `element not visible` error code, so the marshalling of the error between Marionette and geckodriver is broken. Whenever webdriver encounters an error code it does not know about, it falls back to an `unknown error`.

**Comment by andreastt** [comment]
To be clear, this won’t be fixed until the webdriver crate is updated before the 0.16 release.

**Comment by barancev** [comment]
Closed as fixed (in 0.16)? Or won't be fixed?

**Comment by andreastt** [comment]
It’s made it into the webdriver crate, webdriver’s been released, and the dependencies have been bumped on master’s `HEAD`.
